### PR TITLE
Transfer API

### PR DESCRIPTION
**Pull request by dadiorchen**
Features:
  * APIs for token transfer, most of the major function for this part:
    * Transfer with explicit tokens & transfer token with bundle range.
    * Transfer token based on trust or untrust wallets relationships.
    * Two transfer directions: pending & requested.
    * Accept/declient/delete pending/requested transfer transaction.
    * The recode of transfer in the `transaction` table.
    * Wallet can do transfer/accept/decline for sub wallets.

  * A base class for the repositories (BaseRepository), so we can re-use the common CRUD operation. Just like the CRUDRepository in Loopback. 
 
Screen Shot 2020-10-14 at 5 53 59 PM
Screen Shot 2020-10-14 at 5 54 11 PM

### CODE REVIEW

**Review by marcoboncoraglio — APPROVED** [review]
I'm still trying to wrap my head around the business logic, looks correct to me, maybe someone else should also look over. That being said, I'd have a few questions/remarks:

1) Really nice job on abstracting out repository behavior, this everything a lot nicer to read

2) Is there a reason why we need the models to be created by passing a JSON or id? If we have the json object, we could just call the constructor like: `const token = new Token(jsonObj.id)`

Maybe one more general remarks that are a bit more general: Do we have a formatting/linting policy? Do you use eslint or prettier or something like that?

File: server/routes/transferRouter.spec.js [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    sinon.stub(WalletService.prototype, "getById").resolves({
+      transferBundle: async () => {
+        throw new HttpError(202);
+      },
+    });
+    const res = await request(app)
+      .post("/")
+      .send({
+        bundle: {
+          bundle_size: 1,
```
Comment (mzyau) [inline server/routes/transferRouter.spec.js:131]: Should we test what will happen with bundle size > 1 and what gets transferred?

**Review by mzyau — COMMENTED** [review]
Nice! I like how the tests trace out the possibilities for transfers between actor and target wallets, and also permissions for approving transfers.

**Comment by dadiorchen** [comment]
> I'm still trying to wrap my head around the business logic, looks correct to me, maybe someone else should also look over it? That being said, I'd have a few questions/remarks:
> 
> 1. Really nice job on abstracting out repository behavior, this everything a lot nicer to read
> 2. Is there a reason why we need the models to be created by passing a JSON or id? If we have the json object, we could just call the constructor like: `const token = new Token(jsonObj.id)`
> 
> Maybe one more general remarks that are a bit more general: Do we have a formatting/linting policy? Do you use eslint or prettier or something like that?

@marcoboncoraglio thanks for your review. 
Yes, @marcoboncoraglio your suggestion is good, I will think about changing to this way. The way currently we are doing actually is a progress of hesitating that if we need to cache the JSON object in a model.
Previously my thought is DO NOT cache the JSON object, because I am worried in this way, it would lead us to build a complicated ORM-similar system. So I just put the primary key (id) into a model, but at last, I compromised on this, the reason is the performance, do-not-cache way will lead to too much DB visits.

It's sure that we need a formating approach, but because we didn't do it at the beginning, and the code is evolving rapidly, we definitely should choose a good point to bring in the lint or pretiteer into this project to let's cooditate in a better way, let's open an issue for this first.

File: __tests__/supertest.js [hunk 2]
```diff
@@ -63,30 +67,202 @@ describe('Route integration', () => {
...
-  it(`[GET /token/${seed.token.uuid}] Should be able to get a token `, async () => {
+  it(`[GET /tokens/${seed.token.uuid}] Should be able to get a token `, async () => {
     const res = await request(server)
-      .get(`/token/${seed.token.uuid}`)
+      .get(`/tokens/${seed.token.uuid}`)
       .set('treetracker-api-key', apiKey)
       .set('Authorization', `Bearer ${token}`);
     expect(res).to.have.property('statusCode', 200);
-    expect(res.body).to.have.property('token').eq(seed.token.uuid);
+    expect(res.body).to.have.property('uuid').eq(seed.token.uuid);
```
[the remark this answers is not in the document]
Comment (dadiorchen) [reply __tests__/supertest.js:76]: No no no, @mzyau all the code concerning /tokens is temporary, please implement it following the spec, I just changed the code here to let this test pass! I need to visit /tokens API in the transfer tests.

File: __tests__/supertest.js [hunk 3]
```diff
@@ -63,30 +67,202 @@ describe('Route integration', () => {
...
+            expect(res.body.transfers[0]).property("state").eq(Transfer.STATE.completed);
+          });
+
+          it(`Token:#${seed.token.id} now should belong to ${seed.walletB.name}`, async () => {
+            const res = await request(server)
+              .get(`/tokens/${seed.token.uuid}`)
+              .set('treetracker-api-key', apiKey)
+              .set('Authorization', `Bearer ${token}`);
+            expect(res).to.have.property('statusCode', 200);
+            expect(res.body.entity_id).eq(seed.walletB.id);
```
[the remark this answers is not in the document]
Comment (dadiorchen) [reply __tests__/supertest.js:161]: I think so too, there are several places we still are using `entity`.

[the remark this answers is not in the document]
Comment (dadiorchen) [reply __tests__/supertest.js:161]: Maybe we should go through all the code and DB migration and change them at some point.

File: server/routes/transferRouter.spec.js (hunk 1 above)
Comment (dadiorchen) [reply server/routes/transferRouter.spec.js:131]: Yes! that makes sense.

**Review by mzyau — APPROVED** [review]

File: __tests__/supertest.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mzyau) [reply __tests__/supertest.js:76]: Got it!

**Pull request merged**